src/Data/Tree/AVL/Internal.hs

```diff
@@ -80,187 +163,201 @@ instance Ord b => Ord (WithBounds b) where
 data MapLayer h k v self
   = MLBranch
-    { _mlHash      :: h
-    , _mlMinKey    :: WithBounds k
-    , _mlCenterKey :: WithBounds k
-    , _mlTilt      :: Tilt
-    , _mlLeft      :: self
-    , _mlRight     :: self
+    { _mlRevision  :: Revision      -- ^ So we can check if node changed and not rehash it
+    , _mlHash      :: Maybe h       -- ^ Hash of the subtree.
+    , _mlMinKey    :: WithBounds k  -- ^ Minimal key used in tree.
+    , _mlCenterKey :: WithBounds k  -- ^ Minimal key of the right subtree.
+    , _mlTilt      :: Tilt          -- ^ Amount of disbalance.
+    , _mlLeft      :: self          -- ^ Right subtree or subtree hash.
+    , _mlRight     :: self          -- ^ Left subtree or subtree hash.
     }
   | MLLeaf
-    { _mlHash    :: h
-    , _mlKey     :: WithBounds k
-    , _mlValue   :: v
-    , _mlNextKey :: WithBounds k
-    , _mlPrevKey :: WithBounds k
+    { _mlRevision :: Revision
+    , _mlHash     :: Maybe h
+    , _mlKey      :: WithBounds k    -- ^ Key of the leaf node.
+    , _mlValue    :: v               -- ^ Value of the leaf node.
+    , _mlNextKey  :: WithBounds k    -- ^ Next key, for [non]existence check.
+    , _mlPrevKey  :: WithBounds k    -- ^ Prev key.
     }
   | MLEmpty
-    { _mlHash      :: h
+    { _mlRevision :: Revision
+    , _mlHash     :: Maybe h
     }
     deriving (Eq, Functor, Foldable, Traversable, Generic)
 
+deriveEq1 ''MapLayer
+deriveOrd1 ''MapLayer
 deriveShow1 ''MapLayer
 
+-- | AVL tree as whole.
 type Map h k v = Free (MapLayer h k v) h
 
+-- | AVL node that was isolated (hashes were put where in places of subtrees).
+type Isolated h k v = MapLayer h k v h
+
+-- | AVL node that was just assigned its hashes. Required for pruning.
+newtype FreshlyRehashed h k v = FreshlyRehashed
+    { getFreshlyRehashed :: Map h k v
+        -- ^ Get rehashed map.
+    }
+
 #if !MIN_VERSION_free(5,0,2)
 deriving instance Generic (Free t a)
 #endif
 
+-------------------------------------------------------------------------------
+-- * Lenses
+-------------------------------------------------------------------------------
+
+-- | Lenses.
 makeLenses ''MapLayer
 
 -------------------------------------------------------------------------------
--- Typeclasses
+-- * Typeclasses
 -------------------------------------------------------------------------------
 
--- | Class, representing an ability for type to be [de]serialised.
-class Serialisable a where
-    serialise   :: a -> ByteString
-    deserialise :: ByteString -> Either String a
+-- | DB monad, capable of retrieving 'isolate'd nodes.
+class KVRetrieve hash node m where
+    retrieve :: hash -> m node
 
--- | Class, representing DB layer, capable of storing 'isolate'd nodes.
---   The 'store' is an idempotent operation.
-class (Serialisable k, Monad m) => KVStoreMonad k m where
-    retrieve :: Serialisable v => k -> m v
-    store    :: Serialisable v => k -> v -> m ()
+-- | DB monad, capable of storing 'isolate'd nodes _an mass_.
```

Comment:
What does "an mass" mean?
Also, italics are made as `/mim/` in haddock.

Reply:
"In mass". It stores big pack of changes.  Just heard that expression somewhere, probably in https://ru.wikipedia.org/wiki/%D0%9F%D0%BE%D0%BD%D0%B5%D0%B4%D0%B5%D0%BB%D1%8C%D0%BD%D0%B8%D0%BA_%D0%BD%D0%B0%D1%87%D0%B8%D0%BD%D0%B0%D0%B5%D1%82%D1%81%D1%8F_%D0%B2_%D1%81%D1%83%D0%B1%D0%B1%D0%BE%D1%82%D1%83 .

Reply:
you probably wanted to say "en masse", but whatever